Docs 2.0 content technical review (#676)
graknlabs/docs: 0001-typedb/01-start/03-quickstart.md

Reviewer: @jamesreprise : Please perform technical review.

---
pageTitle: Quickstart guide
keywords: getting started, typedb, typeql, tutorial, quickstart, overview
longTailKeywords: get started with typedb, typedb tutorial, typedb quickstart, learn typedb
summary: Learn how to create a TypeDB database, load schema and data, perform queries.
toc: false
---

# Quickstart guide

## Overview

This Quickstart guide goes through the step-by-step process of creating and interacting with a TypeDB database designed 
for an [Identity and Access Management](https://en.wikipedia.org/wiki/Identity_management) solution that will be 
described [later](04-iam-schema.md).

<div class="note">
[Important]
We highly recommend completing this guide. Its goal is to prepare an environment for TypeDB exploration and development.
The resulting database (schema and data) will be needed for example queries throughout the documentation.
</div>

## Prerequisites

> This Quickstart guide takes advantage of TypeDB Studio, a desktop GUI tool for interacting with TypeDB. To proceed, 

Reviewer 2: ```suggestion
This Quickstart guide takes advantage of TypeDB Studio, an IDE (integrated development environment) for TypeDB. To proceed, 
```

please install both TypeDB and TypeDB Studio:

- [Install TypeDB](02-installation.md)
- [Install TypeDB Studio](../../02-clients/01-studio.md#get-typedb-studio)

IMPORTANT:
<div class="note">
[Important]
It’s recommended to use TypeDB and TypeDB Studio of the same version number in order to ensure compatibility. If 

> TypeDB Studio doesn't have a version similar to the TypeDB release — use the closest lower version instead.

Reviewer 2: These sentences aren't very hopeful - we should do what we can to avoid a user ever encountering this scenario, but we might need to provide further detail: if TypeDB Studio is absolutely necessary, they may need to downgrade their version of TypeDB for Studio compatibility.

Author: Downgrading an existing DB installation just because the IDE to work with it didn't updated yet - is less than ideal. Looks very bad on us. 

Also the text complies with recommendations I got in Discord AND with our old docs on that matter - https://docs.vaticle.com/docs/studio/overview#version-compatibility.

Reviewer 2: Alright, let's provide more detail for this on a case by case basis.

Author: Sorry, I'm not sure I follow. What else can we provide if we already have direct instructions on what to do AND a separate compatibility table? I can add the link to the table here if you want.

Reviewer 2: As in, if people come to us with a specific scenario we can give them advice. I think adding a link to the compatibility table would be nice as well.

</div>

## Initialize database

### Start TypeDB

Run the following command in a terminal:

<!-- test-ignore -->
```bash

> typedb server

Reviewer 2: Or .bat for Windows users.

Author: Didn't you just recently fix the typedb binary to be able to run with just `typedb server` on Win if we set PATH variable right? The version with fix from you runs from `typedb server` (in a separate window, but still).

Reviewer 2: Hm, I thought it required the .bat extension still. Excellent!

Author: Let's test it after the release. Just to be sure, OK? 
At this moment, my Win machine refuses to start TypeDB correctly, but it initiates the launch (and shows the ASCII graphics I believe).

```

After showing the TypeDB ASCII logo and the bootup completion time, TypeDB is ready for connections.

![TypeDB server bootup message](../../images/quickstart/quickstart-launched-typedb.png)

### Connect to TypeDB

Launch TypeDB Studio, then:

1. Click the [**Connect to TypeDB**] button (right side of the toolbar).
2. Enter `localhost:1729` in the [**Address**] field.
3. Click the [**Connect**] button (the dialog will close after a successful connection).

![Connection Manager Connected](../../images/studio/connection-interface-connected.png)

### Create a project

TypeDB Studio projects allow you to organize and save related queries for future reuse.

![Project Interface empty](../../images/quickstart/studio-projects-empty.png)

To create a new project:

1. Click the [**Open Project**] button in the [**Project**] panel (upper left).
2. Choose a directory for the project files.
3. Click the [**Open**] button.

> The [**Project**] panel will now display the root directory and a nested hidden directory for unsaved files.

Reviewer 2: Might be worth saying the 'hidden' directory is called '.typedb-studio'

Author: I don't see any reason to do that. We don't use the hidden directory for anything here. I only mentioned it to add some context to users on what they see on screen. We don't need its name to do that.

        $p2(permitted-subject: $su, permitted-access: $ac2) isa permission;
    } then {
        $p1 has validity false;
    };

rule automatic-permission-validity:
    when {
        $p isa permission;
        not {
            $p has validity false;
        };
    } then {
        $p has validity true;
    };
    
rule add-view-access:
    when {
        $modify isa action, has action-name "modify_file";
        $view isa action, has action-name "view_file";
        $ac_modify (accessed-object: $obj, valid-action: $modify) isa access;
        $ac_view (accessed-object: $obj, valid-action: $view) isa access;
        (permitted-subject: $subj, permitted-access: $ac_modify) isa permission;
    } then {
        (permitted-subject: $subj, permitted-access: $ac_view) isa permission;
    };
```

### Upload the schema

To execute the TypeQL statements in the opened file and send them as queries:

1. Ensure the [**Session type**] (schema / data) switch (next to the database dropdown) is set to `schema`.
2. Ensure the [**Transaction type**] (write / read) switch is set to `write`.

   ![Select transaction type](../../images/quickstart/studio-select-transaction-type.png)

> 3. Click the green “play” button (middle of the toolbar) to start the transaction.
> 4. Click the “checkmark” button (left of the “play” button) to commit the changes.

Reviewer 2: ```suggestion
3. Click the 'Run Query' button (middle of the toolbar, identified by a 'play' symbol) to start the transaction.
4. Click the 'Commit Transaction' button (left of the 'Run Query' button, identified by a checkmark symbol) to commit the changes.
```

Author: ```suggestion
3. Click the 'Run Query' button (with a 'play' symbol, in the middle of the toolbar) to start the transaction.
4. Click the 'Commit Transaction' button (with a checkmark symbol, left of the 'Run Query' button) to commit the changes.
```

The Types panel will now display the entity, relation, and attribute types within a type hierarchy of the schema.

![Types browser with IAM schema](../../images/quickstart/studio-types-browser-iam.png)

Data can now be inserted.

## First important queries

### Read the schema

TypeQL can be used to query the schema.

To execute a simple schema query:

1. Ensure the [**Session type**] (schema / data) switch is set to `schema` (next to the database dropdown).
2. Ensure the [**Transaction type**] (write / read) switch is set to `read`.
3. Click the `+` icon right from the [**Project**] panel next to the `iam-schema.tql` tab in the [**Text-editor**] panel.
4. Copy the TypeQL statement below.
5. Click the green “play” button.

<!-- test-ignore -->
```typeql
match $t sub thing;
```
The above query returns all types in the schema and displays the results as a graph.

<div class="note">
[Warning]
The `thing` base type will be deprecated in TypeDB version `3.0`. Consider using `entity`, `attribute`, or `relation` 
base type instead. To produce the same result as the above example, use the following query:

<!-- test-ignore -->
```typeql

> match $s sub $t; { $t type entity; } or { $t type relation; } or { $t type attribute; };

Reviewer 2: This example really does not look very good when I render it in Studio. The structural elegance of our ecosystem might be lost on someone in the face of a graph that looks like this:

<img width="1231" alt="image" src="https://user-images.githubusercontent.com/51956016/231453386-0fa0cd86-5caa-4873-8e65-e9975ef166cd.png">

Reviewer 2: This needs further discussion obviously.

Author: Yes, I do believe that it needs further discussion. Probably out of scope for this review tho.  If you want, I can create an issue on the Studio repo.

insert $p (permitted-subject: $s, permitted-access: $ac) isa permission;

match $s isa subject, has full-name "Pearle Goodman"; $o isa object, has path "budget_2021-08-01.xlsx"; 
      $a isa action, has action-name "modify_file"; $ac (accessed-object: $o, valid-action: $a) isa access; 
insert $p (permitted-subject: $s, permitted-access: $ac) isa permission;

match $s isa subject, has full-name "Pearle Goodman"; $o isa object, has path "LICENSE"; 
      $a isa action, has action-name "modify_file"; $ac (accessed-object: $o, valid-action: $a) isa access; 
insert $p (permitted-subject: $s, permitted-access: $ac) isa permission;

match $s isa subject, has full-name "Pearle Goodman"; $o isa object, has path "README.md"; 
      $a isa action, has action-name "modify_file"; $ac (accessed-object: $o, valid-action: $a) isa access; 
insert $p (permitted-subject: $s, permitted-access: $ac) isa permission;
```

To execute the TypeQL statements copied from code block above:

1. Ensure the [Session type] (schema / data) switch (next to the database dropdown) is set to `data`.
2. Ensure the [Transaction type] (write / read) switch is set to `write`.
3. Click the green “play” button.

> 4. Click the “checkmark” button.

Reviewer 2: Perhaps we should be specifying the instructions to execute above the data rather than below? This would need to be propagated to all other examples.

Author: I don't see a problem here. We are now moved on further and the code block we copied from is above now. The only below I'm seeing here is much higher, before the code-block, so it is correct to say below there. 

Whether the instructions below or above the code block is irrelevant as long as they are correct. 

If you mean consistency problem - we did the very same with schema, but it was divided in two sections for other reasons.

The transaction has been committed, and data can now be queried.

### Read data

To retrieve data from a database:

1. Click the `+` icon right from the [**Project**] panel next to the `iam-data.tql` tab in the [**Text-editor**] panel.
2. Ensure the [Session type] (schema / data) switch (next to the database dropdown) is set to `data`.
3. Ensure the [Transaction type] (write / read) switch is set to `read`.
4. Replace the TypeQL statement in the [**Text-editor**] panel with the one below.
5. Click the green “play” button.

<!-- test-ignore -->
```typeql

> match $f isa file, has path $fp;

Reviewer 2: Great example.
